src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -712,20 +722,65 @@ public virtual unsafe int FindFirstCharacterToEncodeUtf8(ReadOnlySpan<byte> utf8
                 int idx = 0;
 
 #if NETCOREAPP
-                if (Sse2.IsSupported)
+                if (Sse2.IsSupported && utf8Text.Length - 16 >= idx)
                 {
+                    // Hoist these outside the loop, as the JIT won't do it.
+                    Vector128<sbyte> bitMaskLookupAsciiNeedsEscaping = _bitMaskLookupAsciiNeedsEscaping;
+                    Vector128<sbyte> bitPosLookup = Ssse3Helper.s_bitPosLookup;
+                    Vector128<sbyte> nibbleMaskSByte = Ssse3Helper.s_nibbleMaskSByte;
+                    Vector128<sbyte> nullMaskSByte = Ssse3Helper.s_nullMaskSByte;
+
                     sbyte* startingAddress = (sbyte*)ptr;
-                    while (utf8Text.Length - 16 >= idx)
+                    do
                     {
                         Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + utf8Text.Length - 16));
 
                         // Load the next 16 bytes.
                         Vector128<sbyte> sourceValue = Sse2.LoadVector128(startingAddress);
 
-                        Vector128<sbyte> mask = Sse2Helper.CreateAsciiMask(sourceValue);
-                        int index = Sse2.MoveMask(mask);
+                        // Check for ASCII text. Any byte that's not in the ASCII range will already be negative when
+                        // casted to signed byte.
+                        int index = Sse2.MoveMask(sourceValue);
+
+                        if (index == 0)
+                        {
+                            // All of the following 16 bytes is ASCII.
+
+                            if (Ssse3.IsSupported)
+                            {
+                                Vector128<sbyte> mask = Ssse3Helper.CreateEscapingMask(sourceValue, bitMaskLookupAsciiNeedsEscaping, bitPosLookup, nibbleMaskSByte, nullMaskSByte);
+                                index = Sse2.MoveMask(mask);
+
+                                if (index != 0)
+                                {
+                                    idx += BitHelper.GetIndexOfFirstNeedToEscape(index);
+                                    goto Return;
+                                }
+                            }
+                            else
+                            {
+                                byte* p = (byte*)startingAddress;
+                                if (DoesAsciiNeedEncoding(p[0])) goto Return;
+                                if (DoesAsciiNeedEncoding(p[1])) goto Return1;
```

Comment:
These multiple returns break the dependency-chain on `idx` -- and it's inspired from `SpanHelpers<T>.IndexOf`.

Although I have no strong opinion whether to keep it or not.
With SSSE3 available this won't be executed anyway, but I kept it for other platforms where SSSE3 isn't available.